### PR TITLE
[FLINK-22994][Table SQL / Planner] improve the performace of invoking…

### PR DESCRIPTION
**Pull request by zicat**
… nesting udf

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by JingsongLi** [comment]
I just left comments in the jira: https://issues.apache.org/jira/browse/FLINK-22994
FYI

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 1]
```diff
@@ -147,7 +146,21 @@ object BridgingFunctionGenUtil {
         externalOperands,
         returnType)
     } else {
-      generateScalarFunctionCall(ctx, functionTerm, externalOperands, outputDataType)
+      val targetOperands = externalOperands.zip(operands).map{
+        case(targetOperand, sourceOperand) => {
+          if (sourceOperand.isInstanceOf[ExternalGeneratedExpression]) {
+            val expression = sourceOperand.asInstanceOf[ExternalGeneratedExpression]
+            ExternalGeneratedExpression.fromGeneratedExpression(
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:153]: Why do this?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 2]
```diff
@@ -288,16 +302,38 @@ object BridgingFunctionGenUtil {
       s"($externalResultTypeTerm) (${typeTerm(externalResultClassBoxed)})"
     }
     val externalResultTerm = ctx.addReusableLocalVariable(externalResultTypeTerm, "externalResult")
+    val externalCode =
+      s"""
+         |${externalOperands.zip(argumentDataTypes)
+            .map{case (element, dataType) => {
+              if ( isInternal(dataType) || !element.isInstanceOf[ExternalGeneratedExpression]) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:309]: Use match case to gen code and result term.

Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:309]: checkstyle `if ( isInternal `

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 3]
```diff
@@ -288,16 +302,38 @@ object BridgingFunctionGenUtil {
...
+                element.code
+              } else {
+                element.asInstanceOf[ExternalGeneratedExpression].getExternalCode
+              }
+            }}.mkString("\n")}
+         |$externalResultTerm = $externalResultCasting $functionTerm
+         |  .$SCALAR_EVAL(${externalOperands.zip(argumentDataTypes)
+              .map{case (element, dataType) => {
+                if ( isInternal(dataType) || !element.isInstanceOf[ExternalGeneratedExpression]) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:318]: Please all have `|`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExternalGeneratedExpression.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * @param internalTerm term to access the internal result of the expression
+ * @param externalTerm term to access the external result of the expression
+ * @param nullTerm boolean term that indicates if expression is null
+ * @param internalCode code necessary to produce internalTerm and nullTerm
+ * @param externalCode code necessary to produce externalTerm
+ * @param literalValue None if the expression is not literal. Otherwise it represent the
+ *                     original object of the literal.
+ *
+ */
+class ExternalGeneratedExpression(
```
[the remark this answers is not in the document]
Comment (zicat) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExternalGeneratedExpression.scala:36]: One case class extends another case class is not supported in scala 2.11 && 2.12.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala (hunk 1 above)
Comment (zicat) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala]: ExternalGeneratedExpression is not a case class, so the method copy of parent class GeneratedExpression create a new object is the instance of  GeneratedExpression not ExternalGeneratedExpression

Comment (zicat) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala]: I have rollback and override GeneratedExpression copy method in ExternalGeneratedExpression

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 5]
```diff
@@ -288,16 +290,39 @@ object BridgingFunctionGenUtil {
       s"($externalResultTypeTerm) (${typeTerm(externalResultClassBoxed)})"
     }
     val externalResultTerm = ctx.addReusableLocalVariable(externalResultTypeTerm, "externalResult")
+    val externalCode =
+      s"""
+         |${externalOperands.zip(argumentDataTypes)
+            .map{case (element, dataType) => {
+              element match {
+                case external: ExternalGeneratedExpression
+                  if !isInternal(dataType) && !element.resultType.isInstanceOf[StructuredType]
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:299]: Why `isInstanceOf[StructuredType]`?
Why not just `dataType  == element.dataType`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExternalGeneratedExpression.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  extends GeneratedExpression(
+    internalTerm, nullTerm, internalCode, dataType.getLogicalType, literalValue) {
+
+  def getExternalCode: String = externalCode
+
+  def getExternalTerm: String = externalTerm
+
+  def getDataType: DataType = dataType
+
+  override def copy(resultTerm: String, nullTerm: String,
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExternalGeneratedExpression.scala:54]: Why override copy?

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExternalGeneratedExpression.scala:54]: I think it is dangerous, we don't know what changed.
If we can meet the requirements, we can lose the external information.

**Comment by JingsongLi** [comment]
Hi @zicat What do you think about the `copy` method?

**Comment by JingsongLi** [comment]
Can you add a nesting udf ITCase?

**Review by zicat — COMMENTED** [review]
> Can you add a nesting udf ITCase?
Added, Please help to review again.

**Comment by zicat** [comment]
> Hi @zicat What do you think about the `copy` method?

Override copy method is the simplest way for this issue. Exactly It's dangerous when Copy the field of resultType because that maybe cause this.dataType.getLogicalType != super.resultType, So I reinforce the copy method. Copy the other fields is safe I think. Looking forward to your reply.

**Comment by JingsongLi** [comment]
Hi @zicat , it seems that we only have one problem: override copy.
I still feel it is dangerous, not only resultType, but also nullTerm, resultTerm and etc... It may brings a wrong result.

I have tested `SELECT func(func(func(text))) FROM MyTable`. There is no `copy` invoking. Why not just remove `copy`? Let `ExternalGeneratedExpression` fallback to `GeneratedExpression` when `copy` occurred.

Or show your case that `copy` occurred. See can we optimize it.

**Comment by zicat** [comment]
> Hi @zicat , it seems that we only have one problem: override copy.
> I still feel it is dangerous, not only resultType, but also nullTerm, resultTerm and etc... It may brings a wrong result.
> 
> I have tested `SELECT func(func(func(text))) FROM MyTable`. There is no `copy` invoking. Why not just remove `copy`? Let `ExternalGeneratedExpression` fallback to `GeneratedExpression` when `copy` occurred.
> 
> Or show your case that `copy` occurred. See can we optimize it.

You can debug the newest test case of flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtilTest.java, It will invoke the ExternalGeneratedExpression copy method.
Your testing I guess is the batch sql not streaming sql.
Why I override copy method? Because If not override this method, The instance of ExternalGeneratedExpression invoke copy method will return the instance of GeneratedExpression that cause flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala#298 match fail.

**Comment by JingsongLi** [comment]
@zicat Can you create ITCase instead of UT `BridgingFunctionGenUtilTest`? You can create case in `CalcITCase`.

**Comment by zicat** [comment]
> @zicat Can you create ITCase instead of UT `BridgingFunctionGenUtilTest`? You can create case in `CalcITCase`.

Hi @JingsongLi  I add the test function testOptimizeNestingInvokeScalarFunction in `CalcITCase` and delete the UT `BridgingFunctionGenUtilTest`. Please help to review it again, thx.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 7]
```diff
@@ -147,7 +147,8 @@ object BridgingFunctionGenUtil {
         externalOperands,
         returnType)
     } else {
-      generateScalarFunctionCall(ctx, functionTerm, externalOperands, outputDataType)
+      generateScalarFunctionCall(ctx, functionTerm, externalOperands,
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:150]: I think the right way to remove `copy` is in above `prepareExternalOperands`.
- We should check `ExternalGeneratedExpression` in `genToExternalConverterAll`.
- Then remove `case external: ExternalGeneratedExpression` in `generateScalarFunctionCall`, we have already generated the `externalOperands`!

Comment (zicat) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala]: I quite agree with you, I modify the code and please help to review it

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 8]
```diff
@@ -308,7 +315,12 @@ object BridgingFunctionGenUtil {
     operands
       .zip(argumentDataTypes)
       .map { case (operand, dataType) =>
-        operand.copy(resultTerm = genToExternalConverterAll(ctx, dataType, operand))
+        operand match {
+          case external: ExternalGeneratedExpression
+            if !isInternal(dataType) && external.getDataType == dataType
+          => operand.copy(resultTerm = external.getExternalTerm, code = external.getExternalCode)
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:321]: checkstyle: Indentation should be increased

**Review by JingsongLi — APPROVED** [review]
+1, Thanks for the continuous improvement.
- left final comment for checkstyle.
- Can you rebase latest master to resolve conflicts? After resolving, you can do a force push.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala (hunk 8 above)
Comment (zicat) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala]: @JingsongLi  Master code changes a lot, I try to resolve conflicts, Please help to review again.

**Comment by JingsongLi** [comment]
I think this PR is ready to merge. Merging...

**Pull request merged**